tsdb/db.go

```diff
@@ -783,6 +783,17 @@ func (db *DB) Compact() (err error) {
 			db.metrics.compactionsFailed.Inc()
 		}
 	}()
+
+	walTruncationTime := int64(math.MinInt64)
+	walTruncationAttempted, headCompacted := false, false
+	defer func() {
+		if headCompacted && !walTruncationAttempted {
```

Comment:
Not a fan of this. This logic with truncating something in memory then here, then also when head is compacted is kind of hard to follow. Let me look if I can suggest something better quickly.